subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/dependencysubstitution/DefaultDependencySubstitutionsSpec.groovy

```diff
@@ -166,7 +166,7 @@ class DefaultDependencySubstitutionsSpec extends Specification {
     def "cannot substitute with unversioned module selector"() {
         when:
         with(substitutions) {
-            substitute project("foo") with module('group:name')
+            substitute project("foo") using module('group:name')
```

Comment:
@leonard84 - I wonder if I'm running into an issue with Spock combining `with` and `thrown` here.  

This test is [failing for me](https://ge.gradle.org/s/hjwixm4ywcvwq/tests/:dependency-management:test/org.gradle.api.internal.artifacts.ivyservice.dependencysubstitution.DefaultDependencySubstitutionsSpec/cannot%20substitute%20with%20unversioned%20module%20selector?top-execution=1) and [on CI ](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Platform_3_bucket6/55336455?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true)since the thrown exception is no longer a `org.gradle.api.InvalidUserDataException` but instead a `org.spockframework.runtime.ConditionFailedWithExceptionError` caused by the `IUDE` the test is expecting.

If you look at the changes to the test, the prior `with` method was [just forwarding](https://github.com/gradle/gradle/pull/21662/files#diff-fe07a52c38dbd9e35e088aa3b95c1aaa970fdf81366b1b241a1c290ee599202bL271) to the `using` method - so I'm not sure why this would suddenly stop working.  

Yet if I replace the `with` block with the equivalent expanded line:
`substitutions.substitute(substitutions.project("foo")).using(substitutions.module('group:name'))` the test suddenly passes.

I peeked at the Spock issue tracker, but didn't see anything related.  Any ideas?

Reply:
Statements in `with` blocks are treated as implicit assertions, if the top-level expression is a method call, then two things can happen:
1. If it is a void method, then no special treatment is applied
2. If it has a return value, then groovy truth is applied to it, and it is treated similar to `methodCall() == true` with the ` == true` being omitted.

The `Substitution.with` was void while `Substitution.using` returns a substitution, that is why it is now treated differently.

I would suggest to use Groovy's `Object.with` instead of the `Specification.with`, i.e.

```groovy
substitutions.with {
    substitute module(matchingModule) using matchingSubstitute
    substitute module(nonMatchingModule) using nonMatchingSubstitute
}
```
especially when using it outside of the `then`/`expect` blocks, as you don't really want to assert anything here. Alternatively `.tap` could also be used, but since we don't really care about the return value it doesn't really matter here.